Use std OnceCell and LazyLock instead of lazy_static

Qkessler opened the pull request:
Hi there!

Here's my first contribution on the repo, and wanted to take the chance to take a small task that would need doing eventually (if not, for consistency with the rest of the Rust world): Migrating off `lazy_static` to the std once_cell, included in Rust 1.70.0 ([docs](https://doc.rust-lang.org/std/cell/index.html)).

I wrote about this contribution while I investigated the code base on this post, in case you are interested about the investigation and code decisions: https://enriquekesslerm.com/blog/bye-lazy-static-welcome-once-cell/

This pull request also opens the discussion about the inclusion of Nightly features on the code base or not. Depending on where we will take this code base, I believe it's worth considering, or just sticking to stable. The discussion opens up with the two commits I sent:

1. On the first commit, I start small with the `buffer.rs` and `data.rs` static: BUFFERS and FEATURES. On that commit, I don't use anything nightly, and as you'll see, we have to create a helper method to avoid including `get_or_init` on each of the BUFFERS or FEATURES calls.
2. On the second commit, I move on to build.rs, slightly more complex (great, had the chance to learn about the custom build we run on the repo) and used the [LazyLock API](https://doc.rust-lang.org/stable/std/sync/struct.LazyLock.html), still nightly: gated behind the `lazy_cell` feature flag.

Let's continue the conversation directly on the Pull Request! Happy to modify the code with any feedback from the Pull Request, although I'll have limited availability throughout the week (probably be able to respond the most during the weekends).

Best,
Quique.

Qkessler commented on src/main.rs:
```diff
@@ -1,3 +1,4 @@
+#![feature(lazy_cell)]
```
Here's the inclusion on `main.rs` of the unstable feature, in case we want to do this. Let's continue the conversation on the pull request itself!

Qkessler reviewed (commented):
Adding comments on my own PR, try to make the intent clearer.

Qkessler commented:
Also, let's take the chance to think about the CICD process we would love to have on this repo, as currently, there are no checks with the pull requests.

To test this, I ran:
- `cargo +nightly build --release`
- `cargo test`

CeleritasCelery commented:
Also in your blog post you were looking into `build.rs`. We used to not use this, but instead relied on a [complex set of macros](https://github.com/CeleritasCelery/rune/blob/3c8abe90261878cce56a112635bd631c28ae671e/src/macros.rs) to [collect all the definitions in a file](https://github.com/CeleritasCelery/rune/blob/3c8abe90261878cce56a112635bd631c28ae671e/src/editfns.rs#L64) and then [initialize them](https://github.com/CeleritasCelery/rune/blob/3c8abe90261878cce56a112635bd631c28ae671e/src/core/env.rs#L226-L291). Using build.rs to generate them allowed us to make the code [less complex](https://github.com/CeleritasCelery/rune/commit/5cabaf4f7278f47f76237f76bbd12042ae507f38).

Qkessler commented:
> As far as CI/CD goes, I have not looked into setting something up yet. I think your suggestion is a good one, namely running the tests, clippy, and rustfmt. I am not as focused on code coverage because ultimately I would like to have a lot testing run against GNU Emacs as a reference.

This sounds good, and it's the mental framework I would also consider for this project, nice to see we are aligned.

I'll look more into the `build.rs` references, thanks for that!

> As you mentioned in your post, the biggest issue with OnceCell and OnceLock is they are not automatically initialized on first use. LazyCell would solve this, but is not stable yet. Though we have another issue in that we [rely on](https://github.com/CeleritasCelery/rune/blob/fd95f58c7917445098c113cf5f0f79f5351db434/src/main.rs#L46-L48) the lazy_static initialization of INTERNED_SYMBOLS to set the symbols function cells. So we might be able to replace that with [OnceLock](https://doc.rust-lang.org/std/sync/struct.OnceLock.html) since we are initializing it manually anyways.

Got it, like the discussion. Let's keep stable, wanted to have this PR for that decision. With that, I'll move forward with the helper functions for the initialization of the OnceLocks. For the `INTERNED_SYMBOLS` we might get away with initializing at each place, but I find nicer to just have a helper function (with an associated **TODO** to clean it up once LazyLock stabilizes) that the usages can just call if needed; initialized on first usage.

Qkessler commented:
@CeleritasCelery Created an issue to track the work to move to `LazyLock` once it has been stabilized, and added TODOs on the code pointing to the issue. 

Moved everything (buffer.rs, data.rs and build.rs) to use `OnceLock`, by using a helper function on each of them. I find the code to be fine, the helper function is non-intrusive compared to the static upper-case words (we just call `()` on it).

Let me know if you need anything else.

Best,
Quique.

CeleritasCelery commented:
Thanks!
